File: try_scala/hello_world.scala
// scala hello_world.scala
//
// IntelliJ Setup:
// * Scala works because it has "scala-sdk-2.12.8" in the "Global Libraries"

def modMethod1(i: Int) = i % 2 == 0
def modMethod2(i: Int) = { i % 2 == 0 }
def modMethod3(i: Int): Boolean = i % 2 == 0
def modMethod4(i: Int): Boolean = { i % 2 == 0 }

def modMethod(i: Int) = i % 2 == 0

val list = List.range(1, 10)
println(list.filter(modMethod))

val addThenDouble: (Int, Int) => Int = (x,y) => {
    val a = x + y
    2 * a
}

// implicit approach
val add1 = (x: Int, y: Int) => { x + y }
val add2 = (x: Int, y: Int) => x + y

// explicit approach
val add3: (Int, Int) => Int = (x,y) => { x + y }
val add4: (Int, Int) => Int = (x,y) => x + y

val c = scala.math.cos _

val add5 = add1
add5(1,1)

object Stuff {
  def makeStuff() = println("make stuff")
  def makeStuff2(a: Int) = println(s"make 2 ${a}")
}
val s0 = Stuff.makeStuff _
val s1: () => Unit = Stuff.makeStuff
val s2: (Int) => Unit = Stuff.makeStuff2
val s3 = Stuff.makeStuff2 _
s0()
s1()
s2(1)
s3(1)

//-------------------------------------------------
case class Expector(result: Integer) {
  def pass = () => println("Pass")
  def fail = (e: Integer) => println(s"Fail: expected ${result} to equal ${e}")
  def toEqual(expected: Integer) =
    if(result == expected)
      pass()
    else
      fail(expected)
}

val expect = (a: Integer) => Expector(a)

val it = (name: String, test: () => Unit) => test()

it("passes if equal", () =>
  expect(1).toEqual(1)
)

it("fails if not equal", () =>
  expect(1).toEqual(2)
)

File: try_kotlin/hello_world.kt
// kotlinc hello_world.kt -include-runtime -d output/hello.jar
// java -jar output/hello.jar
// IntelliJ Setup:
// *  Kotlin is special and needs a library entry in the .idea/library folder

fun main(args: Array<String>) {
    println("Hello, World!")

  val printer = { println("hello")}
  fun sum0(x: Int, y: Int): Int { return x + y }
  val sum = { x: Int, y: Int -> x + y }
  val sum1: (Int, Int) -> Int = { x, y -> x + y }
  printer()
  println(sum0(1,1))
  println(sum(1,1))
  println(sum1(1,1))

  it("passes if equal", {
    expect(1).toEqual(1)
  })

  it("fails if not equal", {
    expect(1).toEqual(2)
  })
}

fun double(x: Int): Int {
    return 2 * x
}

class Sample() {
    fun foo() { print("Foo") }
}

val s = Sample()

class Expector(val result: Int) {
  fun pass() { println("Pass") }
  fun fail(e: Int) { println("Fail: expected $result to equal $e")}
  fun toEqual(expected: Int) {
    if(result == expected)
      pass()
    else
      fail(expected)
    }
}


val expect = {a: Int -> Expector(a) }

typealias EmptyFn = () -> Unit

val it = { name: String, test: EmptyFn -> test() }

//-------------------------------------------------

// object Stuff {
//   def makeStuff() = println("make stuff")
//   def makeStuff2(a: Int) = println(s"make 2 ${a}")
// }
// val s0 = Stuff.makeStuff _
// val s1: () => Unit = Stuff.makeStuff
// val s2: (Int) => Unit = Stuff.makeStuff2
// val s3 = Stuff.makeStuff2 _
// s0()
// s1()
// s2(1)
// s3(1)

//-------------------------------------------------

